staging/src/k8s.io/component-base/traces/utils.go

```diff
@@ -26,33 +26,64 @@ import (
 	"go.opentelemetry.io/otel/propagation"
 	"go.opentelemetry.io/otel/sdk/resource"
 	sdktrace "go.opentelemetry.io/otel/sdk/trace"
+	"go.opentelemetry.io/otel/semconv"
 	"go.opentelemetry.io/otel/trace"
 
 	"k8s.io/client-go/transport"
-	"k8s.io/klog/v2"
 )
 
-// NewProvider initializes tracing in the component, and enforces recommended tracing behavior.
-func NewProvider(ctx context.Context, baseSampler sdktrace.Sampler, resourceOpts []resource.Option, opts ...otlpgrpc.Option) trace.TracerProvider {
-	opts = append(opts, otlpgrpc.WithInsecure())
+// InitTracing initialized tracing in the component, and enforces recommended tracing behavior
+func InitTracing(ctx context.Context,
+	traceID string,
+	serviceName string,
+	samplingRate *int32,
+	opts []otlpgrpc.Option,
+) (trace.TracerProvider, error) {
+	resourceOpts := []resource.Option{
+		resource.WithAttributes(
+			semconv.ServiceNameKey.String(serviceName),
+			semconv.ServiceInstanceIDKey.String(traceID),
+		),
+	}
+
 	driver := otlpgrpc.NewDriver(opts...)
 	exporter, err := otlp.NewExporter(ctx, driver)
 	if err != nil {
-		klog.Fatalf("Failed to create OTLP exporter: %v", err)
+		return nil, err
 	}
-
 	res, err := resource.New(ctx, resourceOpts...)
 	if err != nil {
-		klog.Fatalf("Failed to create resource: %v", err)
+		return nil, err
 	}
 
+	// Only emit spans when the kubelet sends a request with a sampled trace
+	sampler := sdktrace.NeverSample()
+	// Or, emit spans for a fraction of transactions
+	if samplingRate != nil && *samplingRate > 0 {
+		sampler = sdktrace.TraceIDRatioBased(float64(*samplingRate) / float64(1000000))
+	}
+	// batch span processor to aggregate spans before export.
 	bsp := sdktrace.NewBatchSpanProcessor(exporter)
-
-	return sdktrace.NewTracerProvider(
-		sdktrace.WithSampler(sdktrace.ParentBased(baseSampler)),
+	tp := sdktrace.NewTracerProvider(
+		sdktrace.WithSampler(sdktrace.ParentBased(sampler)),
 		sdktrace.WithSpanProcessor(bsp),
 		sdktrace.WithResource(res),
 	)
+	return tp, nil
+}
+
+// WithTracing adds tracing to requests if the incoming request is sampled
+func WithTracing(handler http.Handler, tp trace.TracerProvider, serviceName string) http.Handler {
+	opts := []otelhttp.Option{
+		otelhttp.WithPropagators(Propagators()),
+		otelhttp.WithPublicEndpoint(),
```

Comment:
Note that WithPublicEndpoint should cause incoming context to be ignored (e.g. would prevent apiserver -> kubelet traces from being linked).  If the handler is on the authenticated port, and comes after auth is handled in the server stack, we don't need this.

Reply:
I would def not have seen that, thanks, removed